File: src/kotlin/BelgianBeers/src/test/java/org/jetbrains/kotlincsharpdemo/D04_Statistics.kt
package org.jetbrains.kotlincsharpdemo

import org.jetbrains.kotlincsharpdemo.TestData.beerFlow
import org.junit.Test

class D04_Statistics {

  inline fun <T> Iterable<T>.avg(f : T.() -> Double) : Double {
    var sum = 0.0
    var count = 0
    forEach {
      count++
      sum += it.f()
    }
    return if (count == 0) 0.0 else sum / count
  }


  @Test
  fun statistics() {

    /// from beer in repository.GetBeers()
    /// where beer.Brewery != null
    /// group beer by beer.Brewery into beersPerBrewery
    ///         orderby beersPerBrewery.Average(beer => beer.Rating) descending
    /// select beersPerBrewery.Key;
    /// take 10

    beerFlow
            .filter { it.Brewery != null }
            .groupBy { it.Brewery }
            .entries
            .sortedByDescending { it.value.avg { Rating } }
            .mapNotNull { it.key }
            .take(10)
            .forEachIndexed { idx, it ->
              println("${idx.inc().toString().padStart(5)}. ${it.Name}")
            }

  }
}

File: src/csharp/BelgianBeers/BelgianBeers/Tests/D04_Statistics.cs
using System.Linq;
using Xunit;
using Xunit.Abstractions;
using static BelgianBeers.Tests.Utilities.TestData;

namespace BelgianBeers.Tests
{
    [Collection("Demo 4 - Statistics")]
    public class D04_Statistics
    {
        private readonly ITestOutputHelper _outputHelper;

        public D04_Statistics(ITestOutputHelper outputHelper)
        {
            _outputHelper = outputHelper;
        }

        [Fact]
        public void Statistics()
        {
            // Statistics:
            var topRatedBreweries = from beer in BeerFlow
                where beer.Brewery != null
                group beer by beer.Brewery into beersPerBrewery
                let average = beersPerBrewery.Average(beer => beer.Rating)
                orderby average descending
                select (brewery: beersPerBrewery.Key, average);

            foreach (var result in topRatedBreweries.Take(10))
            {
                _outputHelper.WriteLine($"{result.average:P} - {result.brewery.Name}");
            }
        }
    }
}